src/common/hobject.cc

```diff
@@ -355,7 +355,11 @@ int cmp(const hobject_t& l, const hobject_t& r)
   return 0;
 }
 
-
+ghobject_t::operator std::string() const {
```

Comment:
Yes that would make sense. OTOH [this comment](https://github.com/ceph/ceph/pull/22458#discussion_r193863070) suggests to use a `str()` method instead of relying on the implicit `operator std::string()` to avoid surprises. What are your thoughts?

Reply:
i am also for the str() `method`, or at least make the `operator std::string` explicit, like 
```c++
explicit operator std::string() const { 
 /// ....
}
```

. @mogeb have you considered reusing a single trace point everywhere and send the fmt string to it, i.e. use it as an alternative of `syslog(3)` without the `...` parameters. if we are fine with this, we can avoid all these troubles. but the downside of this apporach.

Reply:
@tchaikov sure, that's the very first solution I explored (pre-Cephalocon). I initially changed the implementation of `dout()` to just record a tracepoint. With that approach, none of the actual has to be changed. Another alternative could be to use LTTng's [`tracef()`](https://lttng.org/man/3/tracef/v2.8/) (similar to `printf`), but it also needs/uses strings.

My objective here though is to avoid passing strings around for performance reasons.

Reply:
Right.  I think if the legacy dout lines degenerate into a slow string op, we can focus our efforts on replacing dout calls in the fast path and leave the less frequently-run and slow-path code alone until later (e.g., peering).

Reply:
thanks guys. after re-reading `tracing/tracetool/{README.md,parsetool.py`, now i understand the objective and rationale behind this design.

Reply:
@mogeb i agree with @liewegas  on this. and i think we can have a compromise here by differentiating following two use cases:

1. the logging in critical path: to use `trace_${event_name}(loglevel, subsys, [arg_type, arg_name]...)`, so we don't need to render the values of primitive types in string before passing them to LTTng.
1. the logging in non-critical path: to use `trace(loglevel, subsys, fmt_str, [arg]...)`. we can use `fmt::format()` offered by [libfmt](https://github.com/fmtlib/fmt) in the `pasrsetool.py`. as libfmt is already included by ceph as a submodule. and make `trace()` a template function. which calls `tracepoint(subsys, ${default_event_name}_at_${log_evel})` , assuming all combinations of `${default_event_name}_at_${log_level}` `TRACEPOINT_EVENT`s of different `${subsys}`s and `${log_level}`s combinations have been defined by `parsetool.py`. they only have a single `TP_FIELD`.

in this way, we can have

- a unified logging backend, so developers or users don't need to check different sources for tracing an issue, and
- a flexible front-end, so developers don't need to thinking about the name of a trace event, and the argument type of event payload. they only need to do so when it's needed.

what do you think?

Reply:
I think the clang approach would be great. Sounds good, let's rethink it later as you said.

For having a critical and non-critical paths: just to make sure I understand, the non-critical path uses a formatted string using the arguments' ``operator <<()`` functions, correct? If that's the case, we could change the implementation of ``dout`` to record a tracepoint instead of using ``trace(loglevel, subsys, fmt_str, [arg]...)``. This becomes even easier for the developers as nothing changes: ``dout()`` just takes the string passed to it and ends up calling ``tracepoint()`` at the end. I'd need to figure out the details, but I was able to achieve that at the very beginning. Any thought?

Reply:
> the non-critical path uses a formatted string using the arguments' ``operator <<()`` functions, correct?

yes. at this moment, we can use the argument's `operator<<()`, so minimum change is needed. in short and long term, i'd suggest use libfmt's `fmt::format()` instead, it has better performance than that of `std::iostream`. probably we can expose two interfaces

- for the backward compatibility: `dout(1) << "ceph rock" << 5 << '!' << dendl;
- for the faster one: `trace(bluestore, 1, "{0} rock{1}{2}", "ceph", 5, '!');`
- for the fatest one: `trace_ceph_rocks(1, bluestore, string, who, "ceph", int, s, 5, uint_8, wow, '!');`

so, i'd suggest we should start using the `trace()` or `trace_${event_name}` interface in the new code. and replace the implementaion of `dout(1)` for the existing code.